docs/topics/coroutines.rst

```diff
@@ -37,8 +37,101 @@ hence use coroutine syntax (e.g. ``await``, ``async for``, ``async with``):
+several possible implementations of this.
+
+First, such universal :meth:`process_start_requests` can be an asynchronous
+generator itself, and so it will always convert a normal iterable to an
+asynchronous one. Because a result of a middleware method is passed to the same
+method of the next middleware, it's only possible to mix middlewares with
+synchronous and asynchronous :meth:`process_start_requests` if all synchronous
+ones are called first.
+
+.. autofunction:: scrapy.utils.asyncgen.as_async_generator
+
+Here is an example of a universal middleware using this approach::
+
+    from scrapy.utils.asyncgen import as_async_generator
+
+    class ProcessStartRequestsAsyncGenMiddleware:
+        async def process_start_requests(self, start_requests, spider):
+            async for req in as_async_generator(start_requests):
+                # ... do something with req
+                yield req
+
+If this method includes asynchronous code, that code will work even with
+synchronous :meth:`~scrapy.spiders.Spider.start_requests`.
+
+Another option is to make separate methods for normal and asynchronous
+iterables and choose one at run time::
+
+    from inspect import isasyncgen
+
+    class ProcessStartRequestsAsyncGenMiddleware:
+        def _normal_process_start_requests(self, start_requests, spider):
+            # ... do something with normal start_requests
+
+        async def _async_process_start_requests(self, start_requests, spider):
+            # ... do something with async start_requests
+
+        def process_start_requests(self, start_requests, spider):
+            if isasyncgen(start_requests):
+                return self._async_process_start_requests(start_requests, spider)
+            else:
+                return self._normal_process_start_requests(start_requests, spider)
+
+
+.. _async-start_requests-queue:
+
+Queue behavior with asynchronous start_requests
+===============================================
+
+.. versionadded:: 2.x
+
+By default the requests produced by
+:meth:`~scrapy.spiders.Spider.start_requests` don't follow the usual request
+priorities: they are only scheduled while the downloader queue is empty. In
+practice this often means that after some initial requests are scheduled (the
+number depends on the spider concurrency settings), only the requests produced
+by callbacks are scheduled. This behavior is inconvenient, confusing and
+requires workarounds (such as calling ``self.crawler.engine.schedule(request)``
+directly), but changing it would be backwards-incompatible.
+
+A better behavior, which should treat requests from
+:meth:`~scrapy.spiders.Spider.start_requests` in the same way as other
+requests, is enabled when :meth:`~scrapy.spiders.Spider.start_requests` is an
+async function (declared using ``async def``). It doesn't need to contain
+``await`` for this to work, so if you want the new queue behavior, you can just
+change ``def`` to ``async def``.
```

Comment:
I know that's quite hard to act on feedback about docs :) It is also highly subjective. But I think the explanation is not 100% clear here.

This all has nothing to do with request priorities (as request.priority). It is more about if all requests from start_requests are consumed or not, or to which point they are consumed. What do you think about explaining it from this perspective? I would also start with describing the "good" behavior, and then explani the old, historical one. 

Something like this: 

Requests produced by `async def start_requests` are scheduled first. If there is any `await` inside, Scrapy may switch to processing callbacks, but eventually `async def start_requests` resumes its execution as usual (of course, depending on what you're awaiting). 

Note that for historical reasons behavior of non-async, regular `def start_requests` is different: requests are only scheduled while the scheduler queue is empty. In practive this means that only first few requests from start_requests are scheduled (up to MAX_CONCURRENT_REQUESTS), and then processing of start_request stops until all requests produced by callbacks are processed. This behavior is confusing, and may require workarounds; we might change it in future for regular start_requests as well. 

If you want the new queue behavior, you can just change ``def start_requests`` to ``async def start_requests``. Note that in this case all spider(?) middlewares with process_start_requests methods should support async generators. Built-in Scrapy middlewares should work, some third-party or internal middlewares might need to be updated, as described here (todo: add a link).

Reply:
I know my explanations are not 100% clear, as I tried to document things I don't understand myself.

I'll implement your suggestions.